Add walk information (#44)
NUbots/NUbook: src/book/02-system/01-foundations/05-mathematics.mdx

- [Wolfram MathWorld](http://mathworld.wolfram.com/)

- [Eigen Documentation](http://eigen.tuxfamily.org/dox/)

# Interpolation

> Lets say we have a collection of data points. We want to find a function that contains all these points. We call this function an interpolant. Often we use a polynomial to describe this function.

Reviewer: ```suggestion
Let's say we have a collection of data points. We want to find a function that contains all these points. We call this function an interpolant. Often we use a polynomial to describe this function.
```

![2D graph plot of a piecewise polynomial function that passes through specific highlighted points.](./images/maths-spline.svg 'Spline. Interpolation points are in blue.')

Every two consecutive points on the graph have an associated polynomial that is plotted in the range between these two points.

The Quintic Walk specifically uses quintic splines. This means that each polynomial is of the form

$$
S_i = a_{5i} x^5 + a_{4i} x^4 + a_{3i} x^3 + a_{2i} x^2 + a_{1i} x + a_{0i}
$$

For $i = 0, ..., n-1$, where $n$ is the number of interpolation points.

These particular quintic splines have interpolation points $(x_0, y_0, y'_0, y''_0) \cdots (x_n, y_n, y'_n, y''_n)$. They must satify the conditions

- $S_j(x)$ is a polynomial in $[x_j, x_{j+1}]$ for each $j = 0, \cdots , n-1$
- $S_j(x_j) = y_j$ and $S_j(x_{j+1}) = y_{j+1}$ for each $j = 0, \cdots , n-1$
- $S'_{j+1}(x_{j+1}) = S'_j(x_{j+1}) = y'_{j+1}$ for each $j = 0, \cdots , n-2$
- $S''_{j+1}(x_{j+1}) = S''_j(x_{j+1}) = y''_{j+1}$ for each $j = 0, \cdots , n-2$
- $S'_0(x_0) = b_{0}$ and $S'_{n-1}(x_n) = y'_{n}$

> - $S''_0(x_0) = b_{0}$ and $S''_{n-1}(x_n) = y''_{n}$

Reviewer 2: Might be worth mentioning that these constraints are ensuring that join between subsequent polynomials is smooth?

- [Wolfram MathWorld](http://mathworld.wolfram.com/)

- [Eigen Documentation](http://eigen.tuxfamily.org/dox/)

# Interpolation

> Let's say we have a collection of data points. We want to find a function that contains all these points. We call this function an interpolant. Often we use a polynomial to describe this function.

Reviewer: ```suggestion
Given a set of data points, often we want to find a function that contains all these points, which we call an interpolant. Interpolants are commonly polynomials, because they can approximate complex curves with a few points, with relatively low computational costs. However, interpolant polynomials which are locally accurate for the given data can have oscillation issues at the edges of the interval which is being interpolated on - [Runge's phenomenon](https://en.wikipedia.org/wiki/Runge%27s_phenomenon).
```

## Splines

> Splines are used in the [Quintic Walk Engine](/system/subsystems/motion#walk-engine).
> 
> Rather than have one global polynomial, it can be useful to use piecewise polynomials. These are all defined locally around the interpolation points.

Reviewer: ```suggestion
Rather than have one global polynomial, it can be useful to use piecewise polynomials, connected at their edges. These so-called [**splines**](https://en.wikipedia.org/wiki/Spline_(mathematics)) are all defined locally around the interpolation points, and can avoid Runge's phenomenon. Splines are used in the [Quintic Walk Engine](/system/subsystems/motion#walk-engine).
```

- $S''_{j+1}(x_{j+1}) = S''_j(x_{j+1}) = y''_{j+1}$ for each $j = 0, \cdots , n-2$
- $S'_0(x_0) = b_{0}$ and $S'_{n-1}(x_n) = y'_{n}$
- $S''_0(x_0) = b_{0}$ and $S''_{n-1}(x_n) = y''_{n}$

> These conditions result in smooth joints between each polynomial, meaning the position, velocity and acceleration at those joints is the same for both polynomials. Each data point has an x and y value, as well as first and second derivative value. The conditions above ensure the splines satisfy the given values for each data point.

Reviewer 2: ```suggestion
These conditions result in smooth joints between each polynomial, meaning the position, velocity and acceleration at those joints is the same for both polynomials. Each data point has an `x` and `y` value, as well as first and second derivative value. The conditions above ensure the splines satisfy the given values for each data point.
```

- [Wolfram MathWorld](http://mathworld.wolfram.com/)

- [Eigen Documentation](http://eigen.tuxfamily.org/dox/)

# Interpolation

> Given a set of data points, often we want to find a function that contains all these points, which we call an interpolant. Interpolants are commonly polynomials, because they can approximate complex curves with a few points, with relatively low computational costs. However, interpolant polynomials which are locally accurate for the given data can have oscillation issues at the edges of the interval which is being interpolated on - [Runge's phenomenon](https://en.wikipedia.org/wiki/Runge%27s_phenomenon).

Author: We can mention it. I assumed enough contraints would be included to make the interpolant unique, but I guess that should be explicily stated.

> Given a set of data points, often we want to find a function that contains all these points, which we call an interpolant. Interpolants are commonly polynomials, because they can approximate complex curves with a few points, with relatively low computational costs. However, interpolant polynomials which are locally accurate for the given data can have oscillation issues at the edges of the interval which is being interpolated on - [Runge's phenomenon](https://en.wikipedia.org/wiki/Runge%27s_phenomenon).

Reviewer 2: Aliasing can also happen when you have a very large degree on your polynomial. You can get extra wiggles occurring between interpolation points that may not be intended even though the end points align with the the next polynomial in the spline. I think splines are generally broken up in a way to avoid this anyway, but might be worth highlighting this (feel free to override me here if you think its unnecessary)

> Given a set of data points, often we want to find a function that contains all these points, which we call an interpolant. Interpolants are commonly polynomials, because they can approximate complex curves with a few points, with relatively low computational costs. However, interpolant polynomials which are locally accurate for the given data can have oscillation issues at the edges of the interval which is being interpolated on - [Runge's phenomenon](https://en.wikipedia.org/wiki/Runge%27s_phenomenon).

Author: Yeah I guess if you use a polynomial that has degree higher than you need. I've added in some stuff, let me know if it's what you were going for.

## Splines

Rather than have one global polynomial, it can be useful to use piecewise polynomials, connected at their edges. These so-called [**splines**](<https://en.wikipedia.org/wiki/Spline_(mathematics)>) are all defined locally around the interpolation points, and can avoid Runge's phenomenon. Splines are used in the [Quintic Walk Engine](/system/subsystems/motion#walk-engine).

![2D graph plot of a piecewise polynomial function that passes through specific highlighted points.](./images/maths-spline.svg 'Spline. Interpolation points are in blue.')

> Every two consecutive points on the graph have an associated polynomial that is plotted in the range between these two points.

Reviewer 3: ```suggestion
Every pair of consecutive points on the graph has an associated polynomial that is plotted in the range between those two points.
```
